garage/tf/q_functions/discrete_mlp_q_function.py

```diff
@@ -41,25 +43,62 @@ def __init__(self,
                  output_nonlinearity=None,
                  output_w_init=tf.contrib.layers.xavier_initializer,
                  output_b_init=tf.zeros_initializer,
+                 dueling=False,
                  layer_normalization=False):
+        super().__init__(name)
+
+        self._env_spec = env_spec
+        self._hidden_sizes = hidden_sizes
+        self._hidden_nonlinearity = hidden_nonlinearity
+        self._hidden_w_init = hidden_w_init
+        self._hidden_b_init = hidden_b_init
+        self._output_nonlinearity = output_nonlinearity
+        self._output_w_init = output_w_init
+        self._output_b_init = output_b_init
+        self._dueling = dueling
+        self._layer_normalization = layer_normalization
+
+        self.models = []
         obs_dim = env_spec.observation_space.shape
         action_dim = env_spec.action_space.flat_dim
 
-        self.model = DiscreteMLPModel(
-            output_dim=action_dim,
-            name=name,
-            hidden_sizes=hidden_sizes,
-            hidden_nonlinearity=hidden_nonlinearity,
-            hidden_w_init=hidden_w_init,
-            hidden_b_init=hidden_b_init,
-            output_nonlinearity=output_nonlinearity,
-            output_w_init=output_w_init,
-            output_b_init=output_b_init,
-            layer_normalization=layer_normalization)
+        if not dueling:
+            model = DiscreteMLPModel(
+                output_dim=action_dim,
+                hidden_sizes=hidden_sizes,
+                hidden_nonlinearity=hidden_nonlinearity,
+                hidden_w_init=hidden_w_init,
+                hidden_b_init=hidden_b_init,
+                output_nonlinearity=output_nonlinearity,
+                output_w_init=output_w_init,
+                output_b_init=output_b_init,
+                layer_normalization=layer_normalization)
+        else:
+            model = DiscreteMLPDuelingModel(
+                output_dim=action_dim,
+                hidden_sizes=hidden_sizes,
+                hidden_nonlinearity=hidden_nonlinearity,
+                hidden_w_init=hidden_w_init,
+                hidden_b_init=hidden_b_init,
+                output_nonlinearity=output_nonlinearity,
+                output_w_init=output_w_init,
+                output_b_init=output_b_init,
+                layer_normalization=layer_normalization)
 
+        self.models.append(model)
         obs_ph = tf.placeholder(tf.float32, (None, ) + obs_dim, name="obs")
 
-        self.model.build(obs_ph)
+        with tf.variable_scope(self._variable_scope):
```

Comment:
`with self._variable_scope`

Reply:
`self._variable_scope` is a `VariableScope` object, we have to do 
```
with tf.variable_scope(VariableScopeObject):
```
(see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/variable_scope.py#L1804).
Also `VariableScope` is pickleable, `variable_scope` is not.